src/policy/markcompactspace.rs

```diff
@@ -318,69 +321,126 @@ impl<VM: VMBinding> MarkCompactSpace<VM> {
         Self::is_marked(object)
     }
 
-    pub fn calculate_forwarding_pointer(&self) {
-        let start = self.common.start;
-        let end = self.pr.cursor();
-        let mut to = start;
-
-        let linear_scan =
-            crate::util::linear_scan::ObjectIterator::<VM, MarkCompactObjectSize<VM>, true>::new(
-                start, end,
-            );
-        for obj in linear_scan.filter(|obj| Self::to_be_compacted(*obj)) {
-            let copied_size =
-                VM::VMObjectModel::get_size_when_copied(obj) + Self::HEADER_RESERVED_IN_BYTES;
-            let align = VM::VMObjectModel::get_align_when_copied(obj);
-            let offset = VM::VMObjectModel::get_align_offset_when_copied(obj);
-            to = align_allocation_no_fill::<VM>(to, align, offset);
-            let new_obj = VM::VMObjectModel::get_reference_when_copied_to(
-                obj,
-                to + Self::HEADER_RESERVED_IN_BYTES,
-            );
-
-            Self::store_header_forwarding_pointer(obj, new_obj);
-
-            trace!(
-                "Calculate forward: {} (size when copied = {}) ~> {} (size = {})",
-                obj,
-                VM::VMObjectModel::get_size_when_copied(obj),
-                to,
-                copied_size
-            );
+    /// Linear scan all the live objects in the given memory region
+    fn linear_scan_objects(
+        &self,
+        range: Range<Address>,
+        to_be_compacted_only: bool,
+    ) -> impl Iterator<Item = ObjectReference> {
+        let linear_scan = crate::util::linear_scan::ObjectIterator::<
+            VM,
+            MarkCompactObjectSize<VM>,
+            true,
+        >::new(range.start, range.end);
+        linear_scan.filter(move |obj| !to_be_compacted_only || Self::to_be_compacted(*obj))
```

Comment:
I find the condition hard to understand. Maybe move the `filter` to the caller, and just return the `ObjectIterator` here -- that can make the code more intuitive.

Reply:
Fixed